### PR TITLE
[VIKI] Api updated and fixed all bugs

### PR DESCRIPTION
**Pull request by zackmark29**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [ ] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [ ] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement

---

### Description of your *pull request* and other information

Since viki updated all content to DRM, I found a way to find the non drm content. and I've updated all api and fixed some little bugs.

### CODE REVIEW

**Comment by zackmark29** [comment]
@pukkandan I created new PR with some little fixed. The previous PR is gone when I changed my fork default branch

**Comment by TurtleWilly** [comment]
Subtitle download is still broken, it seems:

```
$ youtube-dl --all-subs --skip-download 'https://www.viki.com/tv/37697c-deliveries-from-the-heart-true-stories-from-mercari'
[viki:channel] 37697c: Downloading channel JSON
[viki:channel] 37697c: Downloading JSON metadata
[download] Downloading playlist: Deliveries from the Heart - True stories from Mercari
[viki:channel] playlist Deliveries from the Heart - True stories from Mercari: Collected 6 videos; downloading 6 of them
[download] Downloading video 1 of 6
[viki] 1179099v: Downloading video JSON
[viki] 1179099v: Downloading video streams JSON
[viki] 1179099v: Checking for another mpd url
[viki] 1179099v: Downloading MPD manifest
[info] 1179099v: Downloading 1 format(s): 0+9
[info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].ar.vtt
WARNING: Unable to download subtitle for "ar": HTTP Error 400: Bad Request
[info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].el.vtt
WARNING: Unable to download subtitle for "el": HTTP Error 400: Bad Request
[info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].en.vtt
WARNING: Unable to download subtitle for "en": HTTP Error 400: Bad Request
[info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].es.vtt
WARNING: Unable to download subtitle for "es": HTTP Error 400: Bad Request
[info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].fr.vtt
WARNING: Unable to download subtitle for "fr": HTTP Error 400: Bad Request
[info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].he.vtt
WARNING: Unable to download subtitle for "he": HTTP Error 400: Bad Request
[info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].it.vtt
WARNING: Unable to download subtitle for "it": HTTP Error 400: Bad Request
[info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].ja.vtt
WARNING: Unable to download subtitle for "ja": HTTP Error 400: Bad Request
[info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].pt.vtt
WARNING: Unable to download subtitle for "pt": HTTP Error 400: Bad Request
[info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].ro.vtt
WARNING: Unable to download subtitle for "ro": HTTP Error 400: Bad Request
[info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].tr.vtt
WARNING: Unable to download subtitle for "tr": HTTP Error 400: Bad Request
[download] Downloading video 2 of 6
[viki] 1179098v: Downloading video JSON
…
```

**Comment by zackmark29** [comment]
> Subtitle download is still broken, it seems:
> 
> ```
> $ youtube-dl --all-subs --skip-download 'https://www.viki.com/tv/37697c-deliveries-from-the-heart-true-stories-from-mercari'
> [viki:channel] 37697c: Downloading channel JSON
> [viki:channel] 37697c: Downloading JSON metadata
> [download] Downloading playlist: Deliveries from the Heart - True stories from Mercari
> [viki:channel] playlist Deliveries from the Heart - True stories from Mercari: Collected 6 videos; downloading 6 of them
> [download] Downloading video 1 of 6
> [viki] 1179099v: Downloading video JSON
> [viki] 1179099v: Downloading video streams JSON
> [viki] 1179099v: Checking for another mpd url
> [viki] 1179099v: Downloading MPD manifest
> [info] 1179099v: Downloading 1 format(s): 0+9
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].ar.vtt
> WARNING: Unable to download subtitle for "ar": HTTP Error 400: Bad Request
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].el.vtt
> WARNING: Unable to download subtitle for "el": HTTP Error 400: Bad Request
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].en.vtt
> WARNING: Unable to download subtitle for "en": HTTP Error 400: Bad Request
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].es.vtt
> WARNING: Unable to download subtitle for "es": HTTP Error 400: Bad Request
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].fr.vtt
> WARNING: Unable to download subtitle for "fr": HTTP Error 400: Bad Request
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].he.vtt
> WARNING: Unable to download subtitle for "he": HTTP Error 400: Bad Request
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].it.vtt
> WARNING: Unable to download subtitle for "it": HTTP Error 400: Bad Request
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].ja.vtt
> WARNING: Unable to download subtitle for "ja": HTTP Error 400: Bad Request
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].pt.vtt
> WARNING: Unable to download subtitle for "pt": HTTP Error 400: Bad Request
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].ro.vtt
> WARNING: Unable to download subtitle for "ro": HTTP Error 400: Bad Request
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].tr.vtt
> WARNING: Unable to download subtitle for "tr": HTTP Error 400: Bad Request
> [download] Downloading video 2 of 6
> [viki] 1179098v: Downloading video JSON
> …
> ```

I already updated that

**Comment by TurtleWilly** [comment]
> I already updated that
> check this https://github.com/zackmark29/yt-dlp/releases/tag/2021-09-07
> 
> ```
> C:\DEV\yt-dlp-master\yt-dlp-master\dist>yt-dlp.exe --all-subs --skip-download https://www.viki.com/tv/37697c-deliveries-from-the-heart-true-stories-from-merca
> [viki:channel] 37697c: Downloading channel JSON
> [viki:channel] 37697c: Downloading JSON metadata
> [download] Downloading playlist: Deliveries from the Heart - True stories from Mercari
> [viki:channel] playlist Deliveries from the Heart - True stories from Mercari: Collected 6 videos; downloading 6 of them
> [download] Downloading video 1 of 6
> [viki] 1179099v: Downloading video JSON
> [viki] 1179099v: Downloading video streams JSON
> [viki] 1179099v: Checking for another mpd url
> [viki] 1179099v: Downloading MPD manifest
> [info] 1179099v: Downloading 1 format(s): 0+9
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].ar.vtt
> [download] Destination: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].ar.vtt
> [download] 100% of 11.91KiB in 00:01
> [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].el.vtt
> [download] Destination: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].el.vtt
> ```

Can't get it to work sadly, I'm seeing this:
```
…
[viki:channel] 37697c: Downloading channel JSON
[viki:channel] 37697c: Downloading JSON metadata
[download] Downloading playlist: Deliveries from the Heart - True stories from Mercari
[viki:channel] playlist Deliveries from the Heart - True stories from Mercari: Collected 6 videos; downloading 6 of them
[download] Downloading video 1 of 6
[viki] 1179099v: Downloading video JSON
[viki] 1179099v: Downloading video streams JSON
[viki] 1179099v: Checking for another mpd url
ERROR: name 'mpd_url' is not defined
[download] Downloading video 2 of 6
[viki] 1179098v: Downloading video JSON
[viki] 1179098v: Downloading video streams JSON
[viki] 1179098v: Checking for another mpd url
ERROR: name 'mpd_url' is not defined
[download] Downloading video 3 of 6
[viki] 1179097v: Downloading video JSON
…
```

It's a different error at least. :)

I tried with the latest release from your fork (2021-07-09) and the direct master checkout. Your above link gave me a 404.

**Comment by zackmark29** [comment]
> > I already updated that
> > check this https://github.com/zackmark29/yt-dlp/releases/tag/2021-09-07
> > ```
> > C:\DEV\yt-dlp-master\yt-dlp-master\dist>yt-dlp.exe --all-subs --skip-download https://www.viki.com/tv/37697c-deliveries-from-the-heart-true-stories-from-merca
> > [viki:channel] 37697c: Downloading channel JSON
> > [viki:channel] 37697c: Downloading JSON metadata
> > [download] Downloading playlist: Deliveries from the Heart - True stories from Mercari
> > [viki:channel] playlist Deliveries from the Heart - True stories from Mercari: Collected 6 videos; downloading 6 of them
> > [download] Downloading video 1 of 6
> > [viki] 1179099v: Downloading video JSON
> > [viki] 1179099v: Downloading video streams JSON
> > [viki] 1179099v: Checking for another mpd url
> > [viki] 1179099v: Downloading MPD manifest
> > [info] 1179099v: Downloading 1 format(s): 0+9
> > [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].ar.vtt
> > [download] Destination: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].ar.vtt
> > [download] 100% of 11.91KiB in 00:01
> > [info] Writing video subtitles to: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].el.vtt
> > [download] Destination: Deliveries from the Heart - True stories from Mercari - Episode 6 [1179099v].el.vtt
> > ```
> 
> Can't get it to work sadly, I'm seeing this:
> 
> ```
> …
> [viki:channel] 37697c: Downloading channel JSON
> [viki:channel] 37697c: Downloading JSON metadata
> [download] Downloading playlist: Deliveries from the Heart - True stories from Mercari
> [viki:channel] playlist Deliveries from the Heart - True stories from Mercari: Collected 6 videos; downloading 6 of them
> [download] Downloading video 1 of 6
> [viki] 1179099v: Downloading video JSON
> [viki] 1179099v: Downloading video streams JSON
> [viki] 1179099v: Checking for another mpd url
> ERROR: name 'mpd_url' is not defined
> [download] Downloading video 2 of 6
> [viki] 1179098v: Downloading video JSON
> [viki] 1179098v: Downloading video streams JSON
> [viki] 1179098v: Checking for another mpd url
> ERROR: name 'mpd_url' is not defined
> [download] Downloading video 3 of 6
> [viki] 1179097v: Downloading video JSON
> …
> ```
> 
> It's a different error at least. :)
> 
> I tried with the latest release from your fork (2021-07-09) and the direct master checkout. Your above link gave me a 404.

oh thanks. Just little type. I'll update it now

**Comment by llacb47** [comment]
@zackmark29 revert that, you put the viki extractor in viu.py

**Comment by zackmark29** [comment]
> @zackmark29 revert that, you put the viki extractor in viu.py

hahaha oh you're right. I was checking the viu if there's need to fix =D

**Comment by TurtleWilly** [comment]
Subtitle extraction works now with latest checkout version. 👍

**Comment by zackmark29** [comment]
> Subtitle extraction works now with latest checkout version. 👍

Great :)

**Comment by pukkandan** [comment]
I've made a lot of changes:
* Fix linter, clean and simply code
* Handle any potential errors non-fataly
* Remove the raw mpd URL from formats
* Use `_api_query` and `_call_api` everywhere
* Download all of a channels videos, not just episodes
* `yield` videos so that just enough JSON pages are downloaded when using `--playlist-end` and similar options
* Add `--extractor-arg vikichannel:video_types` to download certain types of videos only

I have tested with free content, but please confirm that I haven't broken login

**Comment by zackmark29** [comment]
> I've made a lot of changes:
> 
> * Fix linter, clean and simply code
> * Handle any potential errors non-fataly
> * Remove the raw mpd URL from formats
> * Use `_api_query` and `_call_api` everywhere
> * Download all of a channels videos, not just episodes
> * `yield` videos so that just enough JSON pages are downloaded when using `--playlist-end` and similar options
> * Add `--extractor-arg vikichannel:video_types` to download certain types of videos only
> 
> I have tested with free content, but please confirm that I haven't broken login

Thanks! that's great. That's what I'm expecting :) since my goal only is to fix the issue and make it work :)

**Comment by pukkandan** [comment]
@zackmark29 , so have you tested that the current PR head works correctly with premium content? If so, I can merge this

**Comment by zackmark29** [comment]
> @zackmark29 , so have you tested that the current PR head works correctly with premium content? If so, I can merge this

Yes it work with premium contents specially the for rent movies. 
I can give you account if you want to test also

File: yt_dlp/extractor/viki.py [hunk 1]
```diff
@@ -101,9 +94,10 @@ def _check_errors(self, data):
                 if reason == 'geo':
                     self.raise_geo_restricted(msg=message)
                 elif reason == 'paywall':
+                    if try_get(data, lambda x: x['paywallable']['tvod']):
+                        self._raise_error('This video is for rent only or TVOD (Transactional Video On demand)')
```
Comment (zackmark29) [inline yt_dlp/extractor/viki.py:98]: This will show up if they don't have purchase yet the movie. 
But for series/episode there's no for rent videos yet

**Comment by Sipherdrakon** [comment]
First I'd like to say thank you for your work on getting this up and running again. It is looking great, I was wondering though why I get a 400 Error when downloading "https://www.viki.com/videos/217687v-flower-boy-next-door-episode-1" It will grab the sneak previews, but not the series episodes themselves.

__main__.py -F "https://www.viki.com/videos/217687v-flower-boy-next-door-episode-1" --verbose
[debug] Command-line config: ['-F', 'https://www.viki.com/videos/217687v-flower-boy-next-door-episode-1', '--verbose']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, pref cp1252
[debug] yt-dlp version 2021.07.07 (source)
[debug] Plugin Extractors: ['SamplePlugin']
[debug] Git HEAD: ee2f938f2
[debug] Python version 3.9.4 (CPython 64bit) - Windows-10-10.0.18362-SP0
[debug] exe versions: ffmpeg 4.4-full_build-www.gyan.dev, ffprobe 4.4-full_build-www.gyan.dev
[debug] Proxy map: {}
[debug] [viki] Extracting URL: https://www.viki.com/videos/217687v-flower-boy-next-door-episode-1
[viki] 217687v: Downloading video JSON
[viki] 217687v: Downloading video streams JSON
ERROR: Unable to download JSON metadata: HTTP Error 400: Bad Request (caused by <HTTPError 400: 'Bad Request'>); please report this issue on  https://github.com/yt-dlp/yt-dlp . Make sure you are using the latest version; see  https://github.com/yt-dlp/yt-dlp  on how to update. Be sure to call yt-dlp with the --verbose flag and include its complete output.
  File "D:\yt-dlp\yt_dlp\extractor\common.py", line 676, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "D:\yt-dlp\yt_dlp\YoutubeDL.py", line 3087, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "C:\Program Files\Python39\lib\urllib\request.py", line 523, in open
    response = meth(req, response)
  File "C:\Program Files\Python39\lib\urllib\request.py", line 632, in http_response
    response = self.parent.error(
  File "C:\Program Files\Python39\lib\urllib\request.py", line 561, in error
    return self._call_chain(*args)
  File "C:\Program Files\Python39\lib\urllib\request.py", line 494, in _call_chain
    result = func(*args)
  File "C:\Program Files\Python39\lib\urllib\request.py", line 641, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)

**Comment by zackmark29** [comment]
@pukkandan Looks like this is not yet included in the exe version. I also tried and it's still using the old script

**Comment by pukkandan** [comment]
@Sipherdrakon @zackmark29 Update

**Comment by Sipherdrakon** [comment]
> @pukkandan Looks like this is not yet included in the exe version. I also tried and it's still using the old script
> but the manual of course working

So you were able to download using the url I provided?? Because I grabbed the commit straight from here and loaded it into the newest source and it still won't work. 

In fact there are a few seasons that will not work:
https://www.viki.com/tv/22330c-emergency-couple
https://www.viki.com/tv/35734c-devilish-joy
https://www.viki.com/tv/37228c-the-immortal-goddess
https://www.viki.com/tv/37489c-please-dont-date-him

There was one that also was hit or miss with episodes within the series but I can't remember which one that is at the moment.

Don't get me wrong this works for 95% of content on VIKI just sadly not the few that I was looking for. :-(

**Comment by zackmark29** [comment]
> > @pukkandan Looks like this is not yet included in the exe version. I also tried and it's still using the old script
> > but the manual of course working
> 
> So you were able to download using the url I provided?? Because I grabbed the commit straight from here and loaded it into the newest source and it still won't work.
> 
> In fact there are a few seasons that will not work:
> https://www.viki.com/tv/22330c-emergency-couple
> https://www.viki.com/tv/35734c-devilish-joy
> https://www.viki.com/tv/37228c-the-immortal-goddess
> https://www.viki.com/tv/37489c-please-dont-date-him
> 
> There was one that also was hit or miss with episodes within the series but I can't remember which one that is at the moment.
> 
> Don't get me wrong this works for 95% of content on VIKI just sadly not the few that I was looking for. :-(

These url are not supported because it doesn't have non drm content.
If you see kocowa tag it means it's drm only
https://www.viki.com/tv/37228c-the-immortal-goddess
https://www.viki.com/tv/37489c-please-dont-date-him

I tried also from your given urls
The ydl only throwing ' Unable to download JSON metadata: HTTP Error 400: Bad Request'
but in my tool this is the actual error from the server
it means, non drm is not available for that.

image

**Comment by Sipherdrakon** [comment]
Hmm, you know what, I was thinking the same thing when I noticed that kocowa tag just wasn't sure as generally it will still download something when I use --allow-unplayable-formats and it still said bad request. Alright no biggie then looks like it functions as it should. :-)

**Comment by zackmark29** [comment]
> Hmm, you know what, I was thinking the same thing when I noticed that kocowa tag but it wasn't on everything that wouldn't download so I wasn't 100%.

Yeah. so if you still gonna get error, the only reason for that, non drm is not available. 
That's only the method to get non drm content.

**Comment by zackmark29** [comment]
> Hmm, you know what, I was thinking the same thing when I noticed that kocowa tag but it wasn't on everything that wouldn't download so I wasn't 100%.

because kocowa have own website kocowa.com which is only drm content

**Comment by Sipherdrakon** [comment]
Okay cool, well like I said it covers 95% of the whole site and I wouldn't expect it to try and grab drm content so this works exactly as it should thank you for all the work on it. :-)

**Comment by zackmark29** [comment]
> Okay cool, well like I said it covers 95% of the whole site and I wouldn't expect it to try and grab drm content so this works exactly as it should thank you for all the work on it. :-)

No, there's only limited kocowa contents. And kocowa is only have drm and the rest can dl non drm. except on some episode specially old

**Pull request merged**